app/lua53/lnodemcu.c

```diff
@@ -166,47 +183,69 @@ static void platform_s_flash_write(const void *from, lu_int32 to, lu_int32 len)
 #define flush_icache(F)   /* not needed */
 
-#endif
+#endif /* LUA_USE_HOST or LUA_USE_ESP */
 
-//============= ESP and HOST lua_debugbreak() test stubs =====================//
 
-#ifdef DEVELOPMENT_USE_GDB
+//================= NodeMCU LFS header and FLash State types =================//
 /*
- * lua_debugbreak is a stub used by lua_assert() if DEVELOPMENT_USE_GDB is
- * defined.  On the ESP, instead of crashing out with an assert error, this hook
- * starts the GDB remote stub if not already running and then issues a break.
- * The rationale here is that when testing the developer might be using screen /
- * PuTTY to work interactively with the Lua Interpreter via UART0.  However if
- * an assert triggers, then there is the option to exit the interactive session
- * and start the Xtensa remote GDB which will then sync up with the remote GDB
- * client to allow forensics of the error.  On the host it is an stub which can
- * be set as a breakpoint in the gdb debugger.
- */
-extern void gdbstub_init(void);
-extern void gdbstub_redirect_output(int);
+** The LFSHeader uses offsets rather than pointers to avoid 32 vs 64 bit issues
+** during host compilation.  The offsets are in units of lu_int32's and NOT
+** size_t, though clearly any internal pointers are of the size_t for the
+** executing architectures: 4 or 8 byte.  Likewise recources are size_t aligned
+** so LFS regions built for 64-bit execution will have 4-byte alignment packing
+** between resources.
+*/
 
-LUALIB_API void lua_debugbreak(void) {
-#ifdef LUA_USE_HOST
-  /* allows debug backtrace analysis of assert fails */
-  lua_writestring(" lua_debugbreak ", sizeof(" lua_debugbreak ")-1);
-#else
-  static int repeat_entry = 0;
-  if  (repeat_entry == 0) {
-    dbg_printf("Start up the gdb stub if not already started\n");
-    gdbstub_init();
-    gdbstub_redirect_output(1);
-    repeat_entry = 1;
-  }
-  asm("break 0,0" ::);
-#endif
-}
-#endif
+typedef struct LFSHeader LFSHeader;
+
+struct LFSHeader{
+  lu_int32 flash_sig;     /* a standard fingerprint identifying an LFS image */
+  lu_int32 flash_size;    /* Size of LFS image in bytes */
+  lu_int32 seed;          /* random number seed used in LFS */
+  lu_int32 timestamp;     /* timestamp of LFS build */
+  lu_int32 nROuse;        /* number of elements in ROstrt */
+  lu_int32 nROsize;       /* size of ROstrt */
+  lu_int32 oROhash;       /* offset of TString ** ROstrt hash */
+  lu_int32 protoROTable;  /* offset of master ROTable for proto lookup */
+  lu_int32 protoHead;     /* offset of linked list of Protos in LFS */
+  lu_int32 shortTShead;   /* offset of linked list of short TStrings in LFS */
+  lu_int32 longTShead;    /* offset of linked list of long TStrings in LFS */
+  short    nFiles;
+  short    nPVs;
+};
 
-//===================== NodeMCU lua.h API extensions =========================//
 
+typedef struct LFSflashState {
+  lua_State  *L;
+  LoadState  *S;
+  int         pass;               /* 1 or 2; also 0 implies 1 */
```

Comment:
Should we be explicit about integer width here too (and in `nSave` below), for consistency if nothing else?

Reply:
Not really.  This an internal state structure (which is private to `nodemcu.c`.  This pattern just follows Lua coding style which uses `int` in its structures for generic integer.  The reason for the `lu_int32` declarations is that these must be 32bit.